torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -54,6 +56,70 @@ def reset_table(self) -> None:
 kernel_side_table = KernelSideTable()
 
 
+###############################################################################
+# Mutation Tracker
+
+
+@dataclasses.dataclass
+class MutationInfo:
+    mutated: bool = False
+    used_in_unknown: bool = False
+
+
+# Super basic mutation tracking pass that tracks which inputs are used in stores
+# It bails if any of the inputs are used in non tl.load/tl.store positions.
+class MutationTracker(ast.NodeVisitor):
+    def __init__(self, infos) -> None:
+        super().__init__()
+        self.infos = infos
+        self.in_load = False
+        self.in_store = False
+
+    def visit_Name(self, node):
+        if node.id not in self.infos:
+            return
+        if self.in_load:
+            pass
+        elif self.in_store:
+            self.infos[node.id].mutated = True
+        else:
+            self.infos[node.id].used_in_unknown = True
```

Comment:
IIUC, this means if the parameters are used in any call that is not `tl.load`, we're going to claim they are mutated. But then it looks a bit too restrictive to me. Can we extend the current no-mutation behavior of `tl.load` to `tl.*` except `tl.store` (because, as far as I know, from all `tl.*`, only `tl.store` actually writes into the pointer)? E.g., if I add `tl.device_print(x)` to the kernel code, the current version of the PR will declare `x` mutated, right?

Reply:
I’d rather have allow list of what loads rather than assume everything except for store does not mutate. But this can just be added on case by case basis

Reply:
Do you know about any other `tl.*` function that mutates the data behind the tensor pointer argument(s)? Quickly skimming through [the list here](https://triton-lang.org/main/python-api/triton.language.html) makes me believe it's only `tl.store`. And then there are (at the bottom of the list) compiler hint ops and debug ops that could be called with pointer args and will lead to unnecessary mutation flagging with the current code?

I totally agree than all non-`tl` calls should be marked as suspicious. But then I'm wondering how deep does our static analysis go: is it just inside the main Triton kernel function's body or also the (potentially nested) helper functions called from inside it? Do we also dive inside the `tl.*` functions?

Reply:
If you're inclined to have an allow list, let's add one with the compiler hint and debug ops in it? I'll be happy with that at this stage, as long as we don't go inside the `tl.*` functions (and treat helper functions as unknown).

Reply:
For posterity: we've discussed this with @oulgen offline. The AST visitor analysis in this PR considers every statement (in the Triton kernel code) in isolation, hence doesn't go into the body of the `tl.*` and other function calls.